dd-java-agent/instrumentation/apache-camel-2.20/src/main/java/org/apache/camel/opentracing/OpenTracingTracer.java

```diff
@@ -0,0 +1,265 @@
+
+    @Override
+    public void notify(EventObject event) throws Exception {
+
+      try {
+        if (event instanceof ExchangeSendingEvent) {
+          ExchangeSendingEvent ese = (ExchangeSendingEvent) event;
+          SpanDecorator sd = getSpanDecorator(ese.getEndpoint());
+          if (!sd.newSpan()) {
+            return;
+          }
+          Span parent = ActiveSpanManager.getSpan(ese.getExchange());
+          SpanBuilder spanBuilder =
+              tracer
+                  .buildSpan(sd.getOperationName(ese.getExchange(), ese.getEndpoint()))
+                  .withTag(Tags.SPAN_KIND.getKey(), sd.getInitiatorSpanKind());
+          // Temporary workaround to avoid adding 'null' span as a parent
+          if (parent != null) {
+            spanBuilder.asChildOf(parent);
+          }
+          Span span = spanBuilder.start();
+          sd.pre(span, ese.getExchange(), ese.getEndpoint());
+          tracer.inject(
+              span.context(),
+              Format.Builtin.TEXT_MAP,
+              new CamelHeadersInjectAdapter(ese.getExchange().getIn().getHeaders()));
+          ActiveSpanManager.activate(ese.getExchange(), span);
+
+          if (LOG.isTraceEnabled()) {
+            LOG.trace("OpenTracing: start client span=" + span);
+          }
+        } else if (event instanceof ExchangeSentEvent) {
+          ExchangeSentEvent ese = (ExchangeSentEvent) event;
+          SpanDecorator sd = getSpanDecorator(ese.getEndpoint());
+          if (!sd.newSpan()) {
+            return;
+          }
+          Span span = ActiveSpanManager.getSpan(ese.getExchange());
+          if (span != null) {
+            if (LOG.isTraceEnabled()) {
+              LOG.trace("OpenTracing: start client span=" + span);
+            }
+            sd.post(span, ese.getExchange(), ese.getEndpoint());
+            span.finish();
+            ActiveSpanManager.deactivate(ese.getExchange());
+          } else {
+            LOG.warn("OpenTracing: could not find managed span for exchange=" + ese.getExchange());
+          }
+        }
+      } catch (Throwable t) {
+        // This exception is ignored
+        LOG.warn("OpenTracing: Failed to capture tracing data", t);
+      }
+    }
+
+    @Override
+    public boolean isEnabled(EventObject event) {
+      return event instanceof ExchangeSendingEvent || event instanceof ExchangeSentEvent;
+    }
+
+    @Override
+    public String toString() {
+      return "OpenTracingEventNotifier";
+    }
+  }
+
+  private final class OpenTracingRoutePolicy extends RoutePolicySupport {
+
+    OpenTracingRoutePolicy(String routeId) {}
+
+    @Override
+    public void onExchangeBegin(Route route, Exchange exchange) {
+      try {
+
+        SpanDecorator sd = getSpanDecorator(route.getEndpoint());
+        Span span =
+            tracer
+                .buildSpan(sd.getOperationName(exchange, route.getEndpoint()))
+                .asChildOf(
```

Comment:
I think here we first have to check if there is active "our" span already present. If yes, then use it as parent, no remote span extracted from the wire

Reply:
Done.